sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -567,13 +550,46 @@ private[sbtplugin] object ScalaJSPluginInternal {
         log.debug(s"with JSEnv ${env.name}")
 
         val input = jsEnvInput.value
-        val config = baseRunConfig.withLogger(sbtLogger2ToolsLogger(log))
+
+        /* The list of threads that are piping output to System.out and
+         * System.err. This is not an AtomicReference or any other thread-safe
+         * structure because the `onOutputStream` callback is always called
+         * within the thread that calls `env.start()`.
+         */
+        var pipeOutputThreads: List[Thread] = Nil
+
+        /* #4560 Explicitly redirect out/err to System.out/System.err, instead
+         * of relying on `inheritOut` and `inheritErr`, so that streams
+         * installed with `System.setOut` and `System.setErr` are always taken
+         * into account. sbt installs such alternative outputs when it runs in
+         * server mode.
+         */
+        val config = RunConfig()
+          .withLogger(sbtLogger2ToolsLogger(log))
+          .withInheritOut(false)
+          .withInheritErr(false)
+          .withOnOutputStream { (out, err) =>
+            for ((optInput, output) <- List(out -> System.out, err -> System.err)) {
+              for (input <- optInput) {
+                val pipeOutputThread = new PipeOutputThread(input, output)
+                pipeOutputThreads ::= pipeOutputThread
+                pipeOutputThread.start()
+              }
+            }
+          }
 
         val run = env.start(input, config)
 
         enhanceNotInstalledException(resolvedScoped.value, log) {
           Await.result(run.future, Duration.Inf)
         }
+
+        /* Wait for the pipe output threads to be done, to make sure that we
+         * do not finish the `run` task before *all* output has been
+         * transferred to System.out and System.err.
+         */
+        for (pipeOutputThread <- pipeOutputThreads)
+          pipeOutputThread.join()
```

Comment:
Good catch! I also thought that we might need to provide this guarantee.

Reply:
On second read, this also has an error swallowing problem: If the `run` method of the `PipeOutputThread` throws (for whatever reason), we will simply consider the thread "done" here (`join` will not re-throw the exception).

Reply:
I think this is actually the desired behavior. The uncaught exception will be printed to `System.err`, and then the thread will be `join`ed. The task will succeed, but I don't think that's a real issue.

Reply:
>  but I don't think that's a real issue.

Wouldn't this be an issue for automated execution? Primarily CI comes to mind. Say for example a piper thread cannot allocate the memory block.

Reply:
Not being able to allocate the memory block would be a fatal error. Even Futures don't catch that, and rather they will typically let the errors be sent to the console. If that is our failure model, I'm sure we have loads of places in the existing codebase with problems like that.

If you feel we must, I can make sure to rethrow here any exception that arises in the piper threads. However, for the TestAdapter case, I don't see any solution: we don't have any place (I think) where we know that the tests are done, and that we should sync up with the piper threads. And in fact, even with the status quo of using redirectIO, I'm not sure we have any guarantee that stuff sent to the console will always be printed before the tasks ends, do we?

Reply:
Perhaps a stronger argument is that is already better than for Scala/JVM. For Scala/JVM, with `fork := true`, sbt uses unmonitored piping threads to forward stdout and stderr. It doesn't monitor exceptions in them, and does not even try to `join()` them.

Being somehow better than Scala/JVM at managing this seems like a stretch goal to me, and if it complicates our code, we should not aim that high.

The same applies for the testing infrastructure.

Reply:
Sad but true...